### DISCUSSION TITLE
Provide example TYPO3 CMS Instance

### DISCUSSION THREAD

**Discussion by pquerner** [sourcebroker/deployer-extended-typo3#1] [General]
Hey,

I'm new to this deployer stuff and I have trouble understanding on how the final deploy.php file must look like when using your repo here.

Could you setup a small TYPO3 instance (can be empty) where you can actually test the deployer stuff on? 

Or-- how an example deploy.php file might look like? The bare bones.

Thanks!

**Comment by kszymukowicz** [comment]
hi,

the point of this whole deployer-extended-typo3 package is to keep deploy.php file as clean and small as possible so the deployment can be easily upgraded. 

My starting version of deploy.php file looks like this and it fits 80% of all deployments without any change.

```
<?php

namespace Deployer;

require_once(__DIR__ . '/vendor/sourcebroker/deployer-loader/autoload.php');
new \SourceBroker\DeployerExtendedTypo3\Loader();

set('repository', 'git@github.com:example/example.git');

server('live', 'vm-dev.example.com')
    ->user('deploy')
    ->set('deploy_path', '/home/www/base-example/live')
    ->set('public_urls', ['https://base.example.com']);

server('beta', 'vm-dev.example.com')
    ->user('deploy')
    ->set('deploy_path', '/home/www/base-example/beta')
    ->set('public_urls', ['https://beta-base.example.com']);

server('local', 'localhost')
    ->set('deploy_path', getcwd())
    ->set('vhost_nocurrent', true)
    ->set('public_urls', ['https://base.example.loc']);
```

In .env file I have:

```
TYPO3_CONTEXT="Development//Local"
INSTANCE="local"

TYPO3__DB__Connections__Default__dbname="dbname"
TYPO3__DB__Connections__Default__host="localhost"
TYPO3__DB__Connections__Default__port="3306"
TYPO3__DB__Connections__Default__user="user"
TYPO3__DB__Connections__Default__password="pass"

```
.env file is out of git and different on each instance. You need to change INSTANCE="xxx" to the name of server defined in server('xxx',...) of deploy.php. This will allow the database operations to work.

This is basically all. What I have in my TYPO3 is that the database settings from .env file are used in TYPO3 automatically so I do not need to keep it in second place somewhere in TYPO3 code. But I think you can start without it also.

**Comment by pquerner** [comment]
> What I have in my TYPO3 is that the database settings from .env file are used in TYPO3 automatically so I do not need to keep it in second place somewhere in TYPO3 code.

I didnt even know that was possible.

Thanks for your example deploy.php. I am sure that will get me started. That deployer package is really awesome I think, but I am just starting so I am not as experienced yet.

**Comment by pquerner** [comment]
The .env file is then copied over before ```dep deploy``` is executed? In what directory (when none exist yet)

**Comment by kszymukowicz** [comment]
Just run first deploy and when it will fail create .env (it will be already empty in /shared folder).

**Comment by pquerner** [comment]
OK it worked now. I setup the TYPO3 instance (just blank database). When I deploy a new version (ie. I bumped TYPO3 CMS version via composer) TYPO3 then thinks its never been installed before. How can I prevent this?

//Edit

```typo3conf/LocalConfiguration.php``` and ```typo3conf/PackageStates.php``` was missing.
Is it a problem adding it to the shared files?
```
add('shared_files', ['typo3conf/LocalConfiguration.php', 'typo3conf/PackageStates.php']);
```

**Comment by kszymukowicz** [comment]
Thats because typo3conf/PackagesStates.php probably. If its absent TYPO3 thinks like it was not installed.
In my TYPO3 I have typo3conf/PackagesStates.php out of git and its automatically crated by typo3_console with composer.json commands.

Look at the script section in my composer.json.

```
"scripts": {
    "packages-states": [
      "@php ./vendor/bin/typo3cms install:generatepackagestates"
    ],
    "folder-structure": [
      "@php ./vendor/bin/typo3cms install:fixfolderstructure"
    ],
    "ext-setup": [
      "@php && ./vendor/bin/typo3cms install:extensionsetupifpossible"
    ],
    "post-autoload-dump": [
      "@packages-states",
      "@folder-structure",
      "@ext-setup"
    ]
  }
```

**Comment by kszymukowicz** [comment]
If you also want to do it like that then you must modify your composer.json reqs like:

```
 "require": {
    "typo3/cms-core": "~8.7.0",

    "typo3/cms-belog": "~8.7.0",
    "typo3/cms-beuser": "~8.7.0",
    "typo3/cms-context-help": "~8.7.0",
    "typo3/cms-cshmanual": "~8.7.0",
    "typo3/cms-fluid-styled-content": "~8.7.0",
    "typo3/cms-filelist": "~8.7.0",
    "typo3/cms-impexp": "~8.7.0",
    "typo3/cms-info": "~8.7.0",
    "typo3/cms-info-pagetsconfig": "~8.7.0",
    "typo3/cms-lowlevel": "~8.7.0",
    "typo3/cms-reports": "~8.7.0",
    "typo3/cms-rsaauth": "~8.7.0",
    "typo3/cms-rte-ckeditor": "~8.7.0",
    "typo3/cms-recycler": "~8.7.0",
    "typo3/cms-setup": "~8.7.0",
    "typo3/cms-sys-note": "~8.7.0",
    "typo3/cms-tstemplate": "~8.7.0",
    "typo3/cms-viewpage": "~8.7.0",

    "jigal/t3adminer": "~9.0.0",
    "fluidtypo3/vhs": "~4.3.0",
    "dmitryd/typo3-realurl": "~2.3.0",

    "sourcebroker/deployer-extended-typo3": "^6.0.0"
  },
```

**Comment by pquerner** [comment]
```
In my TYPO3 I have typo3conf/PackagesStates.php out of git and its automatically crated by typo3_console with composer.json commands.
```

Meaning you have ```typo3conf/LocalConfiguration.php``` in git?

**Comment by kszymukowicz** [comment]
Sure. I have typo3conf/LocalConfiguration.php in git.
Whats the problem with this ?

**Comment by pquerner** [comment]
Are you able to show your ```.gitignore``` if its not too much trouble and doesnt expose critical information?

**Comment by kszymukowicz** [comment]
Here you are:

```
#
# Ignore temporary files (left by editors)
*~
*.bak
*.swp
.DS_Store

#
# Ignore by common IDEs used directories/files
nbproject
*.idea
*.project
.buildpath
.project
.settings
.TemporaryItems
.webprj

#
# Ignore build stuff
bower_components/*
node_modules/*
.cache
.sass-cache
.session

#
# Ignore composer stuff
/bin/*
/vendor/

#
# Ignore common TYPO3 CMS directories
/typo3temp
/fileadmin
/uploads
/typo3
/index.php
/typo3conf/deprecation_*.log
/typo3conf/ENABLE_INSTALL_TOOL

#
# local exclusions
/deploy.lock
/.dep
/typo3conf/realurl_autoconf.php
/typo3conf/PackageStates.php
/typo3conf/ext/*
!/typo3conf/ext/local/
/.env
```

**Comment by pquerner** [comment]
I dont see ```LocalConfiguration.php``` being excluded. Also, is it really a good idea to have this file in git? Because it holds database credentials.

**Comment by kszymukowicz** [comment]
I keep TYPO3 database creds in `.env` file, not in `LocalConfiguration.php`

**Comment by pquerner** [comment]
Ooops, you already mentioned it. Sorry. 
Yea, I'll take a look at this one next. 

I'm getting there, thanks a lot so far for your help! Let me buy you a beer sometime. ^^

**Comment by pquerner** [comment]
How can install extensions from composer which require database updates? I just installed ext:blog and it wouldnt work right away. First I had to disable/enable the extension.

Something like this perhaps?

```
    "updateschema": [
      @php ./vendor/bin/typo3cms database:updateschema"
    ],
```

**Comment by kszymukowicz** [comment]
BTW: Look that "deployer-extended-typo3" is installing for you "helhum/dotenv-connector".
Thanks to "helhum/dotenv-connector" you have all vars from .env file available at very beginning of TYPO3 (when it reads vendor/autoload.php).

So remove database settings from LocalConfiguration.php and in your AdditionalConfiguration.php file (which I have also in git) you can use getenv('smth') to read values from .env and make database settings.

**Comment by kszymukowicz** [comment]
If you use:

```
"scripts": {
    "packages-states": [
      "@php ./vendor/bin/typo3cms install:generatepackagestates"
    ],
    "folder-structure": [
      "@php ./vendor/bin/typo3cms install:fixfolderstructure"
    ],
    "ext-setup": [
      "@php ./vendor/bin/typo3cms install:extensionsetupifpossible"
    ],
    "post-autoload-dump": [
      "@packages-states",
      "@folder-structure",
      "@ext-setup"
    ]
  }
```

then the `typo3cms install:extensionsetupifpossible` should be responsible to make database update schema. Sometimes it fails however - that's probably some bug of typo3_console.

**Comment by pquerner** [comment]
"Sometimes it fails". Thats a bummer. ;D

**Comment by kszymukowicz** [comment]
I never had time to debug this :) Not so critical as when deploy is done the `database:updateschema` is run as separate task. Look on deploy task.
